File: scripts/apply-limiter-policy-default.mjs
import fs from "fs";
import path from "path";
import { readNdjson } from "../lib/ml-policy-dataset.mjs";
import {
  loadLimiterPolicyModel,
  predictLimiterStrength,
  summarizeRowsForStyle,
  summarizeToFeatureVector
} from "../lib/limiter-policy.mjs";
import { getExpressionDefaultStyle } from "../lib/expression-defaults.mjs";

function parseArgs(argv) {
  const out = {};
  for (let i = 0; i < argv.length; i += 1) {
    const item = argv[i];
    if (!item.startsWith("--")) continue;
    const key = item.slice(2);
    const next = argv[i + 1];
    if (!next || next.startsWith("--")) {
      out[key] = true;
      continue;
    }
    out[key] = next;
    i += 1;
  }
  return out;
}

function clamp(n, min, max) {
  return Math.max(min, Math.min(max, n));
}

function toNum(v, fallback = 0) {
  const n = Number(v);
  return Number.isFinite(n) ? n : fallback;
}

function readDefaults(configPath) {
  if (!fs.existsSync(configPath)) return {};
  try {
    return JSON.parse(fs.readFileSync(configPath, "utf8"));
  } catch {
    return {};
  }
}

function main() {
  const args = parseArgs(process.argv.slice(2));
  const cfgPath = path.resolve(String(args.config || "config/expression/defaults.json"));
  const modelPath = String(args.model || "").trim();
  const styleArg = String(args.style || "").trim();
  const alpha = clamp(toNum(args.alpha, 0.65), 0, 1);
  const feedbackFileArg = String(args["feedback-file"] || "").trim();

  const defaults = readDefaults(cfgPath);
  const style =
    styleArg ||
    String(defaults?.selectedRuntime?.style || "").trim() ||
    getExpressionDefaultStyle("natural");
  const styleKey = String(style || "natural").trim().toLowerCase();
  const fallbackStrength = clamp(
    toNum(defaults?.selectedRuntime?.prosodyLimiter?.strength, 0.64),
    0.45,
    0.9
  );
  const feedbackPath = feedbackFileArg
    ? path.resolve(feedbackFileArg)
    : path.resolve(process.cwd(), "data", "training", "feedback.ndjson");
  const rows = readNdjson(feedbackPath);
  const summary = summarizeRowsForStyle(rows, styleKey);
  const feat = summarizeToFeatureVector(summary);
  const policy = loadLimiterPolicyModel(modelPath || undefined);
  const predicted = predictLimiterStrength(policy, feat, fallbackStrength);
  const blended = clamp((1 - alpha) * fallbackStrength + alpha * predicted, 0.45, 0.9);

  defaults.updatedAt = new Date().toISOString();
  defaults.selectedRuntime = {
    ...(defaults.selectedRuntime || {}),
    style,
    prosodyLimiter: {
      enabled: true,
      strength: Number(blended.toFixed(3))
    }
  };
  defaults.limiterPolicy = {
    updatedAt: defaults.updatedAt,
    style,
    alpha,
    fallbackStrength: Number(fallbackStrength.toFixed(3)),
    predictedStrength: Number(predicted.toFixed(3)),
    blendedStrength: Number(blended.toFixed(3)),
    modelMeta: policy?.meta || null,
    feedbackFile: path.relative(process.cwd(), feedbackPath).replace(/\\/g, "/")
  };
  fs.writeFileSync(cfgPath, JSON.stringify(defaults, null, 2), "utf8");
  console.log(
    `limiter_policy_applied style=${style} blended=${blended.toFixed(3)} predicted=${predicted.toFixed(3)} alpha=${alpha}`
  );
}

main();

File: scripts_py/apply_limiter_policy_default.py
import json
import os
from datetime import datetime, timezone
from pathlib import Path

from common_pipeline import parse_args
from limiter_policy import (
    clamp,
    predict_strength,
    read_ndjson,
    load_model,
    summarize_rows_for_style,
    summary_to_feature,
)


def to_num(v, fallback=0.0):
    try:
        n = float(v)
        if n == n and n not in (float("inf"), float("-inf")):
            return n
    except Exception:
        pass
    return fallback


def main():
    args = parse_args(os.sys.argv[1:])
    cwd = Path(os.getcwd())
    cfg_path = Path(str(args.get("config") or "config/expression/defaults.json")).resolve()
    model_path = Path(str(args.get("model") or "models/limiter-policy-v1-py.json")).resolve()
    style_arg = str(args.get("style") or "").strip()
    alpha = clamp(to_num(args.get("alpha"), 0.65), 0.0, 1.0)
    feedback_path = Path(str(args.get("feedback-file") or "data/training/feedback.ndjson")).resolve()

    defaults = {}
    if cfg_path.exists():
        try:
            with open(cfg_path, "r", encoding="utf-8") as f:
                defaults = json.load(f)
        except Exception:
            defaults = {}
    style = (
        style_arg
        or str(((defaults.get("selectedRuntime") or {}).get("style") or "")).strip()
        or str(((defaults.get("selected") or {}).get("style") or "")).strip()
        or "natural"
    )
    fallback = clamp(to_num(((defaults.get("selectedRuntime") or {}).get("prosodyLimiter") or {}).get("strength"), 0.64), 0.45, 0.9)
    rows = read_ndjson(str(feedback_path))
    summary = summarize_rows_for_style(str(cwd), rows, style=str(style).lower())
    feat = summary_to_feature(summary)
    model = load_model(str(model_path))
    predicted = predict_strength(model, feat, fallback=fallback)
    blended = clamp((1 - alpha) * fallback + alpha * predicted, 0.45, 0.9)

    now = datetime.now(timezone.utc).isoformat()
    defaults["updatedAt"] = now
    rt = defaults.get("selectedRuntime", {})
    rt["style"] = style
    rt["prosodyLimiter"] = {"enabled": True, "strength": round(float(blended), 3)}
    defaults["selectedRuntime"] = rt
    defaults["limiterPolicy"] = {
        "updatedAt": now,
        "style": style,
        "alpha": alpha,
        "fallbackStrength": round(float(fallback), 3),
        "predictedStrength": round(float(predicted), 3),
        "blendedStrength": round(float(blended), 3),
        "modelMeta": (model or {}).get("meta") if isinstance(model, dict) else None,
        "feedbackFile": str(feedback_path.relative_to(cwd)).replace("\\", "/"),
    }
    cfg_path.parent.mkdir(parents=True, exist_ok=True)
    with open(cfg_path, "w", encoding="utf-8") as f:
        json.dump(defaults, f, indent=2, ensure_ascii=False)
    print(
        f"limiter_policy_applied style={style} blended={blended:.3f} predicted={predicted:.3f} alpha={alpha}"
    )


if __name__ == "__main__":
    main()
